pkg/didcomm/protocol/messagepickup/service.go

```diff
@@ -0,0 +1,612 @@
+	// get the connection record for the ID to fetch DID information
+	conn, err := s.getConnection(connectionID)
+	if err != nil {
+		return -1, err
+	}
+
+	// generate message ID
+	msgID := uuid.New().String()
+
+	// register chan for callback processing
+	batchCh := make(chan Batch)
+	s.setBatchCh(msgID, batchCh)
+
+	defer s.setBatchCh(msgID, nil)
+
+	// create request message
+	req := &BatchPickup{
+		Type:      BatchPickupMsgType,
+		ID:        msgID,
+		BatchSize: size,
+	}
+
+	// send message to the router
+	if err := s.outbound.SendToDID(req, conn.MyDID, conn.TheirDID); err != nil {
+		return -1, fmt.Errorf("send route request: %w", err)
+	}
+
+	// callback processing (to make this function look like a sync function)
+	var processed int
+	select {
+	case batchResp := <-batchCh:
+		for _, msg := range batchResp.Messages {
+			err := s.handle(msg)
+			if err != nil {
+				logger.Errorf("error handling batch message %s: %v", msg.ID, err)
+				continue
+			}
+			processed++
+		}
+	// TODO https://github.com/hyperledger/aries-framework-go/issues/1134 configure this timeout at decorator level
+	case <-time.After(updateTimeout):
+		return -1, errors.New("timeout waiting for batch")
+	}
+
+	return processed, nil
+}
+
+func (s *Service) getConnection(routerConnID string) (*connection.Record, error) {
+	conn, err := s.connectionLookup.GetConnectionRecord(routerConnID)
+	if err != nil {
+		if errors.Is(err, storage.ErrDataNotFound) {
+			return nil, ErrConnectionNotFound
+		}
+
+		return nil, fmt.Errorf("fetch connection record from store : %w", err)
+	}
+
+	return conn, nil
+}
+
+func (s *Service) getBatchCh(msgID string) chan Batch {
+	s.batchMapLock.RLock()
+	defer s.batchMapLock.RUnlock()
+
+	return s.batchMap[msgID]
+}
+
+func (s *Service) setBatchCh(msgID string, batchCh chan Batch) {
+	s.batchMapLock.Lock()
+	defer s.batchMapLock.Unlock()
+
+	if batchCh == nil {
+		delete(s.batchMap, msgID)
+	} else {
+		s.batchMap[msgID] = batchCh
+	}
+}
+
+func (s *Service) getStatusCh(msgID string) chan Status {
+	s.statusMapLock.RLock()
+	defer s.statusMapLock.RUnlock()
+
+	return s.statusMap[msgID]
+}
+
+func (s *Service) setStatusCh(msgID string, statusCh chan Status) {
+	s.statusMapLock.Lock()
+	defer s.statusMapLock.Unlock()
+
+	if statusCh == nil {
+		delete(s.statusMap, msgID)
+	} else {
+		s.statusMap[msgID] = statusCh
+	}
+}
+
+func (s *Service) handle(msg *Message) error {
+	d, err := json.Marshal(msg.Message)
+	if err != nil {
+		return fmt.Errorf("failed to marshal msg: %v", err)
+	}
+
+	unpackMsg, err := s.packager.UnpackMessage(d)
+	if err != nil {
+		return fmt.Errorf("failed to unpack msg: %v", err)
+	}
+
+	trans := &decorator.Transport{}
+	err = json.Unmarshal(unpackMsg.Message, trans)
+
+	if err != nil {
+		return fmt.Errorf("unmarshal transport decorator : %v", err)
+	}
+
+	// TODO: what does this do?
+	// if trans != nil && trans.ReturnRoute != nil && trans.ReturnRoute.Value == decorator.TransportReturnRouteAll {
```

Comment:
@rolsonquadras ^

Reply:
This is used to store the the socket connection in pool(implemented only in websocket implementation). When an [edge agents](https://github.com/hyperledger/aries-framework-go/blob/master/docs/didcomm_mediator.md#edge-agents-without-inbound-capability) (without inbound capability opens a websocket connection with router), the router will keep the connection in the pool, so that it can send messages back to edge-agent. This logic is not required here and we can remove this commented code.